pythonforandroid/recipes/python3/patches/fix-ctypes-util-find-library.patch

```diff
@@ -10,9 +10,16 @@ diff --git a/Lib/ctypes/util.py b/Lib/ctypes/util.py
 +if True:
 +    def find_library(name):
 +        # Check the user app libs and system libraries directory:
-+        app_root = os.path.normpath(os.path.abspath('../../'))
-+        lib_search_dirs = [os.path.join(app_root, 'lib'), "/system/lib"]
-+        for lib_dir in lib_search_dirs:
++        from jnius import autoclass
++        activity = autoclass('org.kivy.android.PythonActivity').mActivity
++        package_name = activity.getApplicationContext().getPackageName()
++        manager = activity.getApplicationContext().getPackageManager()
++        manager_class = autoclass("android.content.pm.PackageManager")
++        native_lib_dir = manager.getApplicationInfo(
++            manager_class.GET_SHARED_LIBRARY_FILES
++        ).nativeLibraryDir
++        lib_search_dirs = [native_lib_dir, "/system/lib"]
++        for lib_dir in [l for l in lib_search_dirs if os.path.exists(l)]:
```

Comment:
Out of context, but if you feel like it :grimacing: 
I feel like the code below could be more elegant using something like glob.
Using something like (untested):
```python
glob.glob("lib{}.so".format(name)) + glob.glob("{}.so".format(name)) 
```
Well since this is definitely out of context I'm happy that we don't address it in this PR.
I just wanted your advice

Reply:
hmmm I think it would need to be like that:
```
files = [f for f in (glob.glob(os.path.join(lib_dir, "lib{}.*".format(name))) +
                     glob.glob(os.path.join(lib_dir, "{}.".format(name)))
         if f.endswith(".so")]
if len(files) > 0:
    return files[0]
```
I'm not sure I prefer that, it looks a bit unwieldy to me, and my first instinct reading `glob.glob` was to check the docs for how exactly it works, even though I was familiar with the purpose of the module. So if you just want my personal judgement on what I think of it, I would prefer the current way it is written

Reply:
yes I definitely don't like the way you wrote it above :joy: 
But I think it could be written in a much neat way. For instance using the `endswith()` doesn't make sense here. The point of using glob is to not do that.

Let's skip it for now since it's out of context anyway.

Reply:
Both search patterns allow for an overlap, you can't glob that. A regex would work though, but not sure how much more readable that would be?

Reply:
don't really why my initial suggestion would do that: `glob.glob("lib{}.so".format(name)) + glob.glob("{}.so".format(name)) ` it includes both prefix and suffix

Reply:
Does it? I don't think it does, there isn't even a `*` in that, how exactly would it glob anything?

Reply:
Just to clarify, the current code also matches e.g. `libsdl2.amd64.so` (when you try to load `sdl2.so`)

Reply:
That would return `libfoo.so` and `foo.so` for `name = "foo"` so yes it works. If you need some `*` of course glob can do that too.
No I'm suggesting to skip that ugly:
```
            for filename in os.listdir(lib_dir):
                if filename.endswith('.so') and (
                        filename.startswith("lib" + name + ".") or
                        filename.startswith(name + ".")):
```
I mean yes I would either go for a regex or for a glob. Basically the idea (I haven't tried) is to not only skip these `if and or` but eventually also the for look

Reply:
> yes it works

no it doesn't :joy: respective as for the same result

>  No I'm suggesting to skip that ugly:

I got that, but what you are suggesting simply doesn't do the same thing.

I just checked, there is no regex glob either so that would also get us back to list comprehension at the very least. Is that really better?

Reply:
yes that's what I mean, returning a least, just like your `listdir()` does actually. If the list is not empty you know you have it matching. That way you skip the for loop and the if. I think it's fairly readable comparing to that `if and or` thingy

Reply:
Well I realized it's not matching something like `libsdl2.so.3.2` anyway so a regex makes more sense. But I'm not sure I subscribe to the notion that this is any more readable :joy:

Reply:
@AndreMiras okay I tested it with regex now!

Reply:
resolving this since it uses regexes now with even separate tess :+1: